### DISCUSSION TITLE
how to access buildcontext inside a provider

### DISCUSSION THREAD

**Discussion by meroo36** [rrousselGit/riverpod#2939] [Q&A]
I know its not a good idea to store a global BuildContext inside a provider. But what is the proper way to do such things like: showing snackbars, navigation, overlay stuff etc... 

https://github.com/rrousselGit/riverpod/discussions/1331#discussioncomment-2509825

this was really helpful but it doesnt help me to use with [fluttertoast](https://pub.dev/packages/fluttertoast).

Maybe some of you can enlighten me about how you do you handle toast/snackbar inside providers.

**Comment by noriHanda** [comment]
You could use ref.listen inside build method of Consumer

**Reply by meroo36** [reply]
can you provide an example if its possible? And I want to access BuildContext globally using the ref, not only in the corresponding widget. For example, I want to be able to call a toast method in my dio interceptor etc.

Maybe I am thinking completly wrong. Any suggestion is appreciated.

**Reply by noriHanda** [reply]
Accessing BuildContext globally should be avoided since implementing like that prevents you from separating presentation layer and data layer. 

If you want to detect changes in providers, [ProviderObserver](https://riverpod.dev/docs/concepts/provider_observer) may be for you.

A simple way to show toast is as follows.
```dart
class Example extends ConsumerWidget {
  const Example({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(someProvider, (previous, next) {
      if (previous != next)
        ScaffoldMessenger.of(context)
            .showSnackBar(SnackBar(content: Text('your snackbar')));
    });
    return Container();
  }
}
```

**Reply by meroo36** [reply]
I think this is totally okay. But it also seems to me that I need to implement this to every single widget that I want to show a snackbar. This is what I imagine in order to write less code:



```dart
  var dio = Dio(baseOptions)
    ..interceptors.add(InterceptorsWrapper(
      onError: (e, handler) {
        ref.read(scaffoldMessengerPod).showSnackBar(SnackBar(content: Text('your snackbar')));
        throw BaseException("err");
      },
    ));
```
this is inside dioProvider for example.
